### ISSUE TITLE
RFC: 2.0 API Changes - Swift Concurrency

### ISSUE THREAD

**Issue by AnthonyMDev** [apollographql/apollo-ios#3411]
###  **This RFC is a work in progress. Additions and changes will be made throughout the design process. Changes will be accompanied by a comment indicating what sections have changed.**

# Background

The upcoming release of Swift 6 brings some significant changes to the language. The new structured concurrency model is incompatible with the internal mutable state of the existing Apollo iOS infrastructure. While `@unchecked Sendable` can be used to silence most of the errors the current library faces in Swift 6, many of our data structures are only implicitly thread safe, but allows for unsafe usage in ways that would be difficult to account for and prevent if using `@unchecked Sendable`.

The Apollo iOS team has planned to do a large overhaul of the networking APIs for a 2.0 release in the future. Swift 6 is pushing us to move that up on our roadmap.

# Proposal

In order to properly support Swift structured concurrency and Swift 6, we believe significant breaking changes to the library need to be made. We are hoping to use this opportunity to make some of the other breaking changes to the networking layer that we have been planning and release a 2.0 version for Swift 6 compatibility. Due to the time constraints and urgency of releasing a version alongside the official stable release of Swift 6, **we do not expect this 2.0 version to encompass the entire scope of changes we initially wanted to make.** This will be an iterative (though significant) improvement on the existing code base. It is likely that a 3.0 version will be released in the future with additional breaking changes to provide for additional functionality that is out of scope for the Swift 6 compatible 2.0 release.

## Impact - Breaking Changes

For users who are not building custom interceptors, the impact of the 2.0 migration would primarily involve adopt Swift concurrency in your calling code and updating API calls. How easy this would be is dependent on how your existing code is structured. This is the direction the language is going, and if you are upgrading to Swift 6, most of these changes will be necessary anyways.

For users who are doing advanced networking, the migration could require a bit more work. The 2.0 proposal includes significant changes to the way the `RequestChain`, `ApolloInterceptor`, and `NormalizedCache` work. Anyone who is implementing their own custom versions of any of these are going to need restructure their code and make their implementations thread safe.

Users who are unable to migrate will still be able to use Apollo iOS 1.0 with the [`@preconcurrency import` ](https://www.swift.org/migration/documentation/swift-6-concurrency-migration-guide/commonproblems#Preconcurrency-Import) annotation. This would downgrade the compiler errors into warnings in Swift 6.

### Deployment Target

Apollo iOS 2.0 would drop support for iOS 12 and macOS 10.14. The new minimum deployment targets would be:

- iOS 13.0+
- iPadOS 13.0+
- Mac Catalyst 13.0+
- macOS 10.15+
- tvOS 13.0+
- visionOS 1.0+
- watchOS 6.0+

## `RequestChain` and `ApolloInterceptor`

The existing `ApolloInterceptor` protocol is difficult to work with and exposes a lot mutable state that is not always safe to mutate. In 2.0, we aim to make these more structured and prevent unsafe data mutation. Currently, all steps in the request chain are performed using interceptors that provide the following method:

```swift
func interceptAsync<Operation: GraphQLOperation>(
    chain: any RequestChain,
    request: HTTPRequest<Operation>,
    response: HTTPResponse<Operation>?
  ) async -> Result<GraphQLResult<Operation.Data>, any Error>
```

When constructing a custom `RequestChain`, the user is responsible for the error prone process of injecting each of the required interceptors (cache read/write, network fetch, parsing.) at the correct points in the request chain. This results in a lot of copying of boilerplate code and a lot of room for error. In 2.0, we want to streamline this, by hardwiring the required components into the process and allowing other optional interceptors to be added into the chain at pre-flight and post-flight hooks. Instead of passing the `RequestChain` to the interceptors and having them call `chain.proceedAsync()`, the interceptors will return (or throw) and the request chain will know to move on.

### Pre-flight Interceptor (`GraphQLRequestInterceptor`)

Pre-flight interceptors will not have access to the response (which in the past would just be `nil` for interceptors prior to the network fetch. They will instead only be able to intercept the request. They can perform arbitrary code and either pass the request through, mutate the request and return a new request, or throw an error.

```swift
public protocol GraphQLRequestInterceptor {

  func intercept<Operation: GraphQLOperation>(
    request: GraphQLRequest<Operation>
  ) async throws -> GraphQLRequest<Operation>

}
```

### Post-flight Interceptor (`GraphQLResultInterceptor`)

Post flight interceptors will have access to the `GraphQLResult` and the request that was used to get that result. They can perform arbitrary code and either pass the request through or throw an error.

```swift
public protocol GraphQLResultInterceptor {

  func interceptAsync<Operation: GraphQLOperation>(
    request: GraphQLRequest<Operation>,
    result: GraphQLResult<Operation>
  ) async throws

}
```

### Fixed Steps

Between the pre and post-flight interceptors there will be fixed steps:

- Cache read
  - `CacheReadInterceptor` will be removed. This becomes a private implementation detail.
  - Custom cache reading mechanisms will be supported by providing a custom implementation of `NormalizedCache`
- Network Transport
  - `NetworkFetchInterceptor` will be removed. This will be replaced by a `NetworkTransport` object. 
  - Default implementation will be provided, but allow for custom implementations to be injected.
  - See NetworkTransport section for more details
- Response Parsing
  - Will parse raw response data into a `GraphQLResult`
  - _Further research is needed here to define API_
- Cache write
  - `CacheWriteInterceptor` will be removed. This becomes a private implementation detail.
  - Custom cache writing mechanisms will be supported by providing a custom implementation of `NormalizedCache`

### Retrying

The new interceptors won’t have access to the `RequestChain`, so they can’t call `chain.retry()`. Instead, there will be a `RetryRequestError`. If an interceptor throws this error, the request chain will retry the entire request instead of throwing the error.

## Network Transport

In 1.0, we have a `NetworkTransport` protocol which is inaccurately named. Currently `NetworkTransport` encompasses the entire process of a GraphQL request. The entire `RequestChain` is part of a `RequestChainNetworkTransport` implementation. You pass an `InterceptorProvider` to this network transport to construct the interceptors for each request. This allows you to create your own entirely separate `NetworkTransport` objects that don't use the request chain. However this discards a large portion of what Apollo iOS does for you (unless you stitch all of our components together yourself in a similar way).

In 2.0 we are proposing that the `ApolloClient` uses the interceptor pattern directly and `NetworkTransport` becomes a protocol (with default implementation for HTTP and Websockets) which only is concerned with the sending of a request over the network. To handle multi-part (`@defer` and subscription) responses, which will return results multiple times, the `NetworkTransport` will return an `AsyncStream`. Each time the steam returns a new value, the parsing, cache write, and post-flight interceptors will be called. These can be called multiple times for multi-part responses.

```swift
public protocol NetworkTransport {

  func interceptAsync<Operation: GraphQLOperation>(
    request: GraphQLRequest<Operation>
  ) -> AsyncStream<Result<HTTPResponse<Operation>, any Error>>

}
```

## Normalized Cache

_This section is in progress and requires more research._

The `NormalizedCache` API has been too limited, and we are investigating how to allow for more customization of caching implementations. This will likely mean expanding the protocol to receive more information during loading and writing of data to allow for custom implementations to make better decisions about their behavior. We are looking for feedback on what additional functionality users would like to see enabled by the `NormalizedCache`.

The `NormalizedCache` will become an `AnyActor` protocol, meaning implementations will need to be `actor` types in 2.0. This ensures thread safety and prevents data races if a `NormalizedCache` were to be used with multiple `ApolloStores` (which you probably shouldn't do, but is theoretically possible currently).

# Design Questions

These are questions that are currently undecided about this RFC. Please comment on this issue if you have opinions or concerns.

### #2: Do we need to allow interceptors to be inserted in-between these fixed steps?

Simplifying interceptors to pre and post-flight allows for a simpler API and removes the possibility of errors by putting interceptors in the wrong order. Allowing interceptor hooks in-between each of these steps would greatly complicate the API and make reasoning about the state of a request more difficult. However, this is technically removing functionality that is possible in 1.0. 

Interceptors should be used primarily for actually intercepting and altering the process of a request. Throwing an error (or a `RetryRequestError`) is a valid use case. For simply logging data, we should consider designing an API for an event listener that can be attached to a client or request.

Is there any use case in which it is necessary to intercept:
- A cached response prior to returning a result
- The raw data of a network response prior to parsing
- The parsed result of a network response prior to writing it to the cache
  - Customization of the response data written to the cache can be done with a custom `NormalizedCache`.

**Comment by TizianoCoroneo** [comment]
Hey 👋  I checked the list of interceptors that we use in our project. Most of them would become `GraphQLRequestInterceptor` and fit nicely with the new model.
However, there is one interceptor that I'm not sure how to translate. 

<details><summary>ErrorParsingInterceptor</summary>
<p>
```
public class ErrorParsingInterceptor: ApolloInterceptor {
    public var id: String = "ErrorParsingInterceptor"
    public init() {}

    public func interceptAsync<Operation>(
        chain: RequestChain,
        request: HTTPRequest<Operation>,
        response: HTTPResponse<Operation>?,
        completion: @escaping (Result<GraphQLResult<Operation.Data>, Error>) -> Void
    ) where Operation: GraphQLOperation {
        do {
            guard let response = response else {
                assertionFailure("We need to have a response to parse the GQL errors in the response. This interceptor is placed in the wrong order in the request chain.")
                throw InterceptorOrderError()
            }

            switch response.httpResponse.statusCode {
            case 410:
                throw GraphQLHTTPResponseError(
                    operationName: String(describing: Operation.self),
                    body: response.rawData,
                    response: response.httpResponse,
                    kind: .upgradeNeeded
                )

            case 401:
                throw GraphQLHTTPResponseError(
                    operationName: String(describing: Operation.self),
                    body: response.rawData,
                    response: response.httpResponse,
                    kind: .notAuthorized
                )

            case 503:
                throw GraphQLHTTPResponseError(
                    operationName: String(describing: Operation.self),
                    body: response.rawData,
                    response: response.httpResponse,
                    kind: .maintenance
                )

            case _ where !response.httpResponse.isSuccessful:
                throw GraphQLHTTPResponseError(
                    operationName: String(describing: Operation.self),
                    body: response.rawData,
                    response: response.httpResponse,
                    kind: .errorResponse
                )

            default:
                chain.proceedAsync(
                    request: request,
                    response: response,
                    interceptor: self,
                    completion: completion)
            }

        } catch {
            chain.handleErrorAsync(
                error,
                request: request,
                response: response,
                completion: completion)
        }
    }
}
```
</p>
</details> 

This interceptor has the simple task of translating HTTP error codes into enum cases. I'm not too fond of it (it doesn't add much value), but it's an example of something that doesn't look like it would fit the new model, as we insert this in-between the `NetworkFetchInterceptor` and the `JSONResponseParsingInterceptor`, since in case we receive one of these HTTP codes the JSON payload is not present in the form expected by the parser.
What kind of error we would get back from the framework in case of, for example, a 401 is a bit unclear. 

Another point of uncertainty is the `additionalErrorInterceptor` functionality. We are currently using it to create signpost ranges for network requests, that are sometimes handy when profiling the app. Our interceptor chain looks like this:


<details><summary>ErrorParsingInterceptor</summary>
<p>
```
 override open func interceptors<Operation>(
        for _: Operation
    ) -> [ApolloInterceptor] where Operation: GraphQLOperation {
      [
            SignpostInterceptor(
                type: .begin,
                name: "Request chain",
                message: "Start: %@ - %@"
            ),

            MaxRetryInterceptor(maxRetriesAllowed: 1),
            loggerInterceptor,

            detectMultipleCallsInterceptor,

            SignpostInterceptor(
                type: .event,
                name: "Request chain",
                message: "Cache read start: %@ - %@"
            ),
            CacheReadInterceptor(store: self.store),
            tokenInterceptor,
            metadataInterceptor,
            botProtectionInterceptor,

            SignpostInterceptor(
                type: .event,
                name: "Request chain",
                message: "Networking start: %@ - %@"
            ),

            NetworkFetchInterceptor(client: self.client),

            SignpostInterceptor(
                type: .event,
                name: "Request chain",
                message: "Response parsing start: %@ - %@"
            ),
            errorParsingInterceptor,
            JSONResponseParsingInterceptor(),
            AutomaticPersistedQueryInterceptor(),

            SignpostInterceptor(
                type: .event,
                name: "Request chain",
                message: "Cache write start: %@ - %@"
            ),
            CacheWriteInterceptor(store: self.store),

            SignpostInterceptor(
                type: .end,
                name: "Request chain",
                message: "End: %@ - %@"
            )
      ]
}

    override open func additionalErrorInterceptor<Operation>(
        for _: Operation
    ) -> ApolloErrorInterceptor? where Operation: GraphQLOperation {
        SignpostErrorInterceptor(
            type: .end,
            name: "Request chain",
            nextErrorInterceptor: apiErrorInterceptor
        )
    }
```
</p>
</details> 

We're using the additional error interceptor to terminate the signpost range when an error is thrown. For our use-case, we would like to have either something similar to `additionalErrorInterceptor` to catch errors before they are thrown outside the client, or to have os.log signposts embedded in the client directly.

**Comment by jimisaacs** [comment]
Hello! This is a quick comment as it is all I have time for at the moment, but expect follow-ups and more discussion.

I have concerns about taking away cache read/write interceptors. There's a fundamental difference in providing these versus a custom normalized cache, which is the former operates on GraphQL operations, while the latter operates on cache records. Also the former operates on requests and responses and has all that context, including the request context itself, while the latter does not.

I also have concerns about taking away the APIs that exist in the current NetworkTransport layer, but the reasoning here is more bespoke. We are taking advantage of being able to wrap this entire later in order to allow for a new URL/endpoint if needed upon request, which currently isn't possible with only the request chain. Doing it at this layer allows us to not have to recreate a client for endpoint changes.

For NormalizedCache, we would need a more robust way to be notified of any/all cache activity, as you know already. Though we also need a way to implement cache expiration, which we currently have implemented as a cache read interceptor.

We also have GraphQL request tracing (logging) implemented as an interceptor.

**Comment by AnthonyMDev** [comment]
Thank you both for your feedback @TizianoCoroneo & @jimisaacs. That's really helpful. I'm going to need to do some more thinking and will update this RFC soon.

@jimisaacs 
> I also have concerns about taking away the APIs that exist in the current NetworkTransport layer, but the reasoning here is more bespoke. We are taking advantage of being able to wrap this entire later in order to allow for a new URL/endpoint if needed upon request, which currently isn't possible with only the request chain. Doing it at this layer allows us to not have to recreate a client for endpoint changes.

In 1.0, you can modify the endpoint URL by mutating the `graphQLendpoint` property of the `HTTPRequest` in an interceptor. Would this enable your use case here?

I don't think there is any reason why a `NetworkTransport` in the 2.0 couldn't be wrapped in the same way though!

**Comment by jimisaacs** [comment]
> In 1.0, you can modify the endpoint URL by mutating the graphQLendpoint property of the HTTPRequest in an interceptor. Would this enable your use case here?

It technically would, though some more context, what I have is an injection pattern, where you initialize a client with a config object that basically configures how internal NetworkTransport will be created on every request. This includes passing through an interceptorProvider which is considered a passthrough at this layer. So to remove the current NetworkTransport facility, and to still allow the same functionality, it will require understanding how we configure interceptor providers in this new layer.

Ultimately to move this configuration to an interceptor would mean that particular interceptor is internal only, which would mean that interceptors would need to be appended to base interceptors. Which would require a redesign of all this being passthrough.

**Comment by jimisaacs** [comment]
> What additional functionality would you like to see enabled by the NormalizedCache.
I think I've referenced Apollo Web's [Type Policies](https://www.apollographql.com/docs/react/caching/cache-configuration/#customizing-type-policies) before, but this would be useful. Basically something that could provide that relationship we were looking for in a query for a list of things with a query for a single thing, where cache normalization would be linked between the two.

**Comment by Brett-Best** [comment]
Excited by the proposed changes here, does an approximate timeframe exist for the v2 release?

**Comment by AnthonyMDev** [comment]
Due to the excellent feedback I've gotten here, we've decided to go in a different direction with the 2.0. There will still be significant changes to support `async/await`, but the `RequestChain` format will be staying much closer to the existing design. I'll be updating this RFC with some more accurate information tomorrow.

As for timeframe, we are committed to having at least a beta of 2.0 released alongside the September stable release of Swift 6. I'm hoping we can get the beta out earlier than that and be approaching a stable release version of Apollo iOS 2.0 with Swift 6, but at the very least we will have a working beta for you to begin your migration with.

**Comment by AnthonyMDev** [comment]
The RFC has been updated to reflect the current state of work on 2.0. This is still subject to change and additional feedback is appreciated!

**Comment by JOyo246** [comment]
I haven't seen any talk about giving SelectionSet sendable conformance. Is there a timeline on this, or reasoning why it's not already sendable?

**Comment by AnthonyMDev** [comment]
@JOyo246 I haven't added that part to the RFC yet, because I haven't gotten to working on that part of the library yet. Making these safely `Sendable` was a question that I needed to do some research into, because some of the models are actually mutable (for use in a local cache mutation). And there have been requests to allow them to be mutated outside of cache transactions for other use cases. I believe that, since we are using copy on write semantics, we should be able to allow the mutable models to be `Sendable`, but I wasn't sure about that at the time of beginning this RFC.

I'll update the RFC once I've confirmed.

Once I've determined a solution, if we are able to make the models `Sendable` without introducing breaking changes, I will be backporting `Sendable` conformance to the models in 1.0. So you won't have to wait for 2.0 to use them.

**Comment by swizzlr** [comment]
I don't know if this was covered in other discussion but I just wanted to note here that I believe it's overly restrictive to make the NormalizedCache API (or really, any protocol) AnyActor. A type doesn't need to be an Actor to be Sendable (ie, usable across concurrency domains/threads), which seemed to be the motivation here. An actor would probably be a good choice but that can probably be an implementation detail?

**Comment by CraigSiemens** [comment]
> Instead of passing the RequestChain to the interceptors and having them call chain.proceedAsync(), the interceptors will now return a NextAction (or throw) and the request chain will use that action to proceed onto the next interceptor.
>
> ```swift
>  func intercept<Operation: GraphQLOperation>(
>    request: HTTPRequest<Operation>,
>    response: HTTPResponse<Operation>?
>  ) async throws -> RequestChain.NextAction<Operation>
> ```

One downside I can see with this approach is it doesn't allow an interceptor to hook into completion side of the request. Previously the completion closure passed to `chain.proceedAsync` could have additional logic added to it that could use the result passed to the completion.

For example, here's a simplified version of our logging interceptor. Our actual implementation uses signposts for logging, but the structure is the same. This might only work in our project since we don't use caching so it's guaranteed that the completion will only be called once. 

```swift
class LoggingInterceptor: ApolloInterceptor {
    func interceptAsync<Operation: GraphQLOperation>(
        chain: RequestChain,
        request: HTTPRequest<Operation>,
        response: HTTPResponse<Operation>?,
        completion: @escaping (Result<GraphQLResult<Operation.Data>, Error>) -> Void
    ) {
        print("Starting \(request.operation.operationName)")
        
        chain.proceedAsync(
            request: request,
            response: response
        ) { result in
            switch result {
            case .success
                print("Success \(request.operation.operationName)")
            case .failure
                print("Failure \(request.operation.operationName)")
            }
            
            completion(result)
        }
    }
}
```

I wonder if there's a way to implement it to still allow the interceptor to hook into the request going down the list of interceptors and back up the list on completion. I'm not sure how possible that'd be with caching allowing the interceptors to "return" multiple times. One thought would be to split the interceptors for caching from the ones from the ones doing the network request. That'd allow each list of interceptors to know that only one value will be returned and then a higher level type would responsible for sending both responses to the caller.

---

I've also been using the [`ClientMiddleware` type in `OpenAPIRuntime`](https://github.com/apple/swift-openapi-runtime/blob/main/Sources/OpenAPIRuntime/Interface/ClientTransport.swift#L220) recently. It uses async/await and serves a similar purpose while allowing intercepting the request and response. The flexibility of it has been really nice allowing implementing thing like logging, automatic retry for requests, and some types of cacheing (ex. `returnCacheDataElseFetch`).

They've got a number of [examples in their repo](https://github.com/search?q=repo%3Aapple%2Fswift-openapi-generator+path%3A*Middleware.swift&type=code) but here's a simple logging example with the same behaviour as the `LoggingInterceptor` above.

```swift
func intercept(
    _ request: HTTPRequest,
    body: HTTPBody?,
    baseURL: URL,
    operationID: String,
    next: @Sendable (HTTPRequest, HTTPBody?, URL) async throws -> (HTTPResponse, HTTPBody?)
) async throws -> (HTTPResponse, HTTPBody?) {
    print("Starting \(operationID)")
    do {
        let response = try await next(request, body, baseURL)
        print("Success \(operationID)")
        return response
    } catch {
        print("Failure \(operationID)")
        throw error
    }
}
```

**Comment by jimisaacs** [comment]
+1 to @CraigSiemens `LoggingInterceptor` point. We have a logging interceptor that looks very similar. Would like to know how to achieve the same thing with this RFC.

**Comment by AnthonyMDev** [comment]
Thanks for the input all.

@swizzlr I'll think about this more before moving forward. My intention here was to ensure that cache read/writes are done serially and prevent races conditions. Currently, we are using a barrier queue to handle that in the `ApolloStore` when it calls into your cache implementation. If you were to access the `NormalizedCache` outside of `ApolloStore`, then there are no guarantees of thread safety.

I'd love to hear opinions from others on this one.

---

@CraigSiemens @jimisaacs 

Thanks for providing this use case here. The way I would think to handle that would be to just use two interceptors. A `PreRequestLoggingInterceptor` and a `PostResponseLoggingInterceptor` or really any combination of logging interceptors you need. For the logging use case specifically, another possible way to handle this would be to inject a logger into your `RequestContext`, and then all of your custom interceptors could just log events as necessary.

While something similar to the example @CraigSiemens provided is doable, it does make for a less clean API in my opinion. Though I certainly want to ensure that we aren't preventing any necessary use cases here. Are there other use cases outside of logging in which this would be useful, and that breaking the behavior into two separate interceptors would not be a workable or desirable solution?

**Comment by jimisaacs** [comment]
+1, the `next()` pattern is a well-established part of middleware designs and provides the capability for managing centralized logic, as described. It’s widely used in command pattern-inspired architectures like Flux/Redux, where unidirectional data flow and middleware are employed. Moving away from this pattern and suggesting the splitting of centralized logic into multiple interceptors introduces unnecessary complexity and risks leaking business logic across multiple components, which breaks encapsulation and violates principles like Single Responsibility.

**Comment by AnthonyMDev** [comment]
Thanks for the feedback. I'm going to go back to the drawing board and make sure we get this right!

**Comment by Killectro** [comment]
Hey @AnthonyMDev! It's been ~a month so I figured I'd check in on how progress was going with 2.0?

**Comment by AnthonyMDev** [comment]
Hey there! We're making progress. The suggestions in this RFC have made us re-think the Interceptor APIs, and the Apollo team has been busy preparing for and attending GraphQL Conference this month. I'm hopping back into this today and hope to have an idea of the direction on those APIs and the RFC updated by end of week.

I'm also considering trying to release a public preview of _some_ of this work that isn't breaking in an Alpha release if I'm able to do it without breaking existing APIs. Things like versions of the `ApolloClient.fetch` APIs that return an `AsyncStream` instead of using a completion handler (and maybe `Sendable` for the generated models?). That way we could get some users testing these things while we iterate on the 2.0.

**Comment by woodymelling** [comment]
Hey @AnthonyMDev  

With these upcoming changes, will Apollo's interceptors properly receive [TaskLocal](https://developer.apple.com/documentation/swift/tasklocal) values from the request callsite?

With full structure concurrency, this happens automatically, but with the class-based interceptors, I believe that extra work needs to be done. 

It would be really great if the interceptors are run with the same TaskLocals as the request callsite. 

Libraries such as [swift-dependencies](https://github.com/pointfreeco/swift-dependencies?tab=readme-ov-file) rely on this functioanlity quite heavily

**Comment by AnthonyMDev** [comment]
Looking for more feedback here!

> I have concerns about taking away cache read/write interceptors. There's a fundamental difference in providing these versus a custom normalized cache. The former operates on GraphQL operations, while the latter operates on cache records. Also the former operates on requests and responses and has all that context, including the request context itself, while the latter does not. - @jimisaacs

I understand your point about the cache mechanism needing context on the request/operation, especially until we implement `@typePolicy` & `@fieldPolicy`. I'm feeling like having the caching mechanisms as part of the `RequestChain` is a bad design though, and I'd like to consider changing that.

`CacheReadInterceptor` is the only interceptor that can terminate the `RequestChain` early, and if it does (b/c it got a valid cache hit with a fetchPolicy `.returnCacheDataElseFetch`) then none of the interceptors after it are called. In the new RFC, that is modeled using the `exitEarlyAndEmit` action. I don't really love the existing behavior in 1.0 where your future interceptors aren't even called when the cache read exits early. It is an unexpected behavior that can trip people up. 

It also can emit a cache response and then continue on with the request chain (with a fetchPolicy `.returnCacheDataAndFetch`). That uses the `proceedAndEmit(intermediaryResult:...)` action. I'm also not a fan of this being part of the `RequestChain`, because the intermediaryResult does not travel through the chain on its way out. It's just emitted as is (this is the existing behavior in 1.0).

---

My feeling is that the awkwardness here is because the cache read/write really isn't part of the "networking" process and should instead, be handled outside of the `RequestChain`. I'm thinking that we could expose a secondary `protocol CacheInterceptor` that would also be provided via the `InterceptorProvider`. Instead of having all interceptors in one place, you will provide a list of `requestInterceptors` and optionally, a single `cacheInterceptor`. 

This `CacheInterceptor` would have access to the `GraphQLOperation` with all of its variables. I don't know if we want to expose the `HTTPRequest/HTTPResponse` here, but we could. I could see a situation in which you would want to expose the initial `HTTPRequest` to the `CacheInterceptor`, if you wanted to use a custom URL-based cache or something.

This feels like better separation of concerns. I think it's less confusing when your `networkInterceptors` are never called because there was a cache hit. You either get a request that proceeds through all of them, or none of them are called in the first place.

**Do you think this would be too limiting?** The only direct behavior this would remove is the ability for a `networkInterceptor` to emit an intermediary value. That would only be possible from the cache getting a hit and then continuing on to the request chain. B/c the interceptors will return the `GraphQLResult` using the `next()` middleware pattern, the ability to early exit would still technically exist (by returning a result without calling `next()`), but it would not be recommended usage.

Maybe the issues with the intermediary and early exit results not traveling through the request chain isn't really an issue to anyone?

**Comment by AnthonyMDev** [comment]
@woodymelling 
> With these upcoming changes, will Apollo's interceptors properly receive [TaskLocal](https://developer.apple.com/documentation/swift/tasklocal) values from the request callsite?

> With full structured concurrency, this happens automatically, but with the class-based interceptors, I believe that extra work needs to be done.

I'm not clear what work needs to be done to support this, but I'd like to make it possible. Once I get an initial alpha release out, maybe you can play around with it and let me know what is/isn't working.

**Comment by CraigSiemens** [comment]
I like the idea of separating the cache interceptor from the request interceptors.

> Maybe the issues with the intermediary and early exit results not traveling through the request chain isn't really an issue to anyone?

I can't think of an issue with it. It's currently possible today if an interceptor called `completion` instead of calling `proceedAsync`.

---

:+1: for @cheif's suggestions.

I don't think that would be too difficult. The `CachePolicy` enum could be split up into two based on the number of results that the operation will produce.

```swift
enum SingleResultCachePolicy {
    case returnCacheDataElseFetch
    case fetchIgnoringCacheData
    case fetchIgnoringCacheCompletely
    case returnCacheDataDontFetch
}

enum MultiResultCachePolicy {
    case returnCacheDataAndFetch
}
```

Then there could be two versions of fetch for the different cache policies and return types. 

```swift 
func fetch<Query: GraphQLQuery>(
    query: Query,
    cachePolicy: SingleResultCachePolicy = .default,
    context: (any RequestContext)? = nil
) async throws -> GraphQLResult<Query.Data>
{ ... }

func fetch<Query: GraphQLQuery>(
    query: Query,
    cachePolicy: MultiResultCachePolicy,
    context: (any RequestContext)? = nil
) -> AsyncThrowingStream<GraphQLResult<Query.Data>, any Error>
{ ... }
```

Since the functions have the same label, Xcode's suggestions will contain the cases from both enums.

![Screenshot 2024-09-25 at 9 46 55 AM](https://github.com/user-attachments/assets/a4508d5f-688f-4418-a33c-01b6c0fd216c)

Unless the return type is already matching, in which case it'll only show the matching cases.

![Screenshot 2024-09-25 at 9 55 30 AM](https://github.com/user-attachments/assets/9aa0648e-90f0-47ea-9fc5-ff5dbad80e8b)

![Screenshot 2024-09-25 at 9 54 13 AM](https://github.com/user-attachments/assets/5dac5371-3a94-413d-be6c-7ce098b43613)

**Comment by calvincestari** [comment]
@cheif @CraigSiemens 

> I agree with the overall idea that we might get multiple results, and thus returning an `AsyncThrowingStream` is the correct API. However, this will only happen for one (of five) cache policies, the others will always return exactly once, and thus a simple `async throws` method would suffice for these cases.

This is also applicable to `@defer` and `@stream` operations though, which are independent of any cache policy.

**Comment by cheif** [comment]
> @cheif @CraigSiemens
> 
> > I agree with the overall idea that we might get multiple results, and thus returning an `AsyncThrowingStream` is the correct API. However, this will only happen for one (of five) cache policies, the others will always return exactly once, and thus a simple `async throws` method would suffice for these cases.
> 
> This is also applicable to `@defer` and `@stream` operations though, which are independent of any cache policy.

I see, I haven't come across `@defer` and `@stream` previously, so I don't really understand how they are treated by the client-libs. My naive take would be that maybe the codegen could generate out different `GraphQLQuery` when it comes across `@defer`/`@stream`, and these could then be used for specializing down `.fetch`, e.g.:
```
public protocol SingleResultGraphQLQuery: GraphQLQuery {}
public protocol StreamingGraphQLQuery: GraphQLQuery {}
...
func fetch<Query: SingleResultGraphQLQuery>(..., cachePolicy: SingleResultCachePolicy) async throws -> ...
func fetch<Query: StreamingGraphQLQuery>(..., cachePolicy: SingleResultCachePolicy) -> AsyncThrowingStream<...>
```
etc. 

It would lead to more overloads of `.fetch`, but I would assume that this just happens at the outermost layer of the library, while everything internally could just return `AsyncThrowingStream`.

**Comment by TizianoCoroneo** [comment]
> My naive take would be that maybe the codegen could generate out different GraphQLQuery when it comes across @defer/@stream, and these could then be used for specializing down .fetch, [...]

I agree with this. The library should absolutely offer a simplified async/await API for the base case without defer/stream/returnCacheDataElseFetch.

**Comment by wiedem** [comment]
> I agree with the overall idea that we might get multiple results, and thus returning an `AsyncThrowingStream` is the correct API. However, this will only happen for one (of five) cache policies, the others will always return exactly once, and thus a simple `async throws` method would suffice for these cases.

+1 A strong vote for this from my side.
In the majority of my projects in which I use Apollo-iOS, the vast majority of fetches produce a single result. In my opinion, it should be clear from the API whether a fetch can return multiple results or not. And if it can't, no `AsyncThrowingStream` should be returned.

Anything else leads to confusion as to whether you as a developer have to deal with the fact that there may or may not be multiple results, although it should actually be clear from the usage.

**Comment by AnthonyMDev** [comment]
Thanks for all the feedback on the desire for a single return `async` fetch function. We'll devise some solution for that.

We’ll need to do some exploration to see what shape that could take. I don’t love the idea of overloading, because it’s odd to me that the API of the same function could change in a way that significantly alters how you consume it. But a function with a slightly different name that is used just for single-result fetches is fine with me.

**Comment by drkibitz** [comment]
On our client wrapper around Apollo, we provide:
- fetch: async throws
- fetch: return AsyncThrowingStream
- perform: async throws
- upload: async throws
- watch: return AsyncStream
- subscribe: return AsyncStream

It is up to the consumer to decide which flavor of fetch to use, and it has worked that way so far this year just fine.

**Comment by jimisaacs** [comment]
In our client wrapper around ApolloClient, we provide the following methods:
- fetch: async throws
- fetch: return AsyncThrowingStream
- perform: async throws
- upload: async throws
- watch: return AsyncStream
- subscribe: return AsyncStream

The consumer decides which flavor of fetch to use, and this has worked out just fine this year. This was also thought about extensively, albeit internally.

**Comment by AnthonyMDev** [comment]
@jimisaacs How do users decide which overload of fetch to use? Just by casting/inferring the return type? What if the user chooses to use the `async throws` version for a request that would return multiple results?

**Comment by jimisaacs** [comment]
> How do users decide which overload of fetch to use? Just by casting/inferring the return type?
Yes
> What if the user chooses to use the async throws version for a request that would return multiple results?
It would result in receiving only the first result.

**Comment by AnthonyMDev** [comment]
@CraigSiemens I'm implementing this concept of a separate `cacheInterceptor` and then a list of `requestInterceptors`. But I'm realizing that this may have an impact on logging interceptors. There is a chance that the `requestInterceptors` would never be called. The `cacheInterceptor` would be run first, and if it gets a cache hit with a cache policy of `.returnCacheDataElseFetch` then the `requestInterceptors` are never called. Also, if you use a `.returnCacheDataDontFetch` policy, the `requestInterceptors` are never called.

If you have a request interceptor to do logging, that won't be called prior to cache read, and it won't be called at all if you don't get to the network request portion of the chain.

My thought is that the correct way of handling this going forward would be to inject a logger that you can then use in any of your interceptors. This could be done using a global dependency injection container if you use some sort of dependency injection library; passing a logger to your interceptors during creation in your `InterceptorProvider`; passing it into the `RequestContext` on `fetch`; or possibly via TaskLocal values?

Does this limit any necessary use cases for anyone?

**Comment by AnthonyMDev** [comment]
As I've been building this out, I'm realizing that the middleware design that @CraigSiemens and @jimisaacs have asked for doesn't really work with our networking stack. This is because interceptors may be called multiple times 
when using `@defer`, or when getting a cache hit and then a network response.

With `async/await`, these `next()` closures only return once. The only way to make this work would be to make the `next()` closure itself return an `AsyncThrowingStream`. Custom interceptors would then need to iterate over that stream, and the code becomes very confusing and error-prone very quickly.

I'm exploring an alternative where each interceptor has two functions. One for intercepting the request on the way in and another for intercepting the response on the way out. I hope that will allow you to encapsulate logic within a single interceptor, while still enabling interceptors to be called multiple times for the same request.

**Comment by fabioferrero** [comment]
@AnthonyMDev First, thanks to the work you are doing here. I think is the proper way to proceed with Apollo :)
I came here for the Sendable conformance of generate objects, but I guess we can wait for Apollo 2.0
What is the planned roadmap at this point? We want to place Apollo 2.0 release in our roadmap to Swift 6 complete concurrency migration

**Comment by AnthonyMDev** [comment]
I'm working hard on getting this ready, but it's a huge undertaking. I can't guarantee when it will be done yet, but I'll be continuing to post updates here in the mean time.

Until then, plan to import Apollo using `@preconcurrency` as all of our current implementations should be thread safe. 

I'm hoping to have a significant update by the end of the week. Just trying to get things actually building so I can verify that what I've done so far works. There were so many areas that needed to be updates, getting to a point where the code even compiles again has been a challenge. :)

**Comment by AnthonyMDev** [comment]
Unfortunately still marching forward with the work, but nothing that is functional yet. I've been making a lot of design iterations to accommodate all the feedback. Will continue to update here as I have news to share.

**Comment by woodymelling** [comment]
This will not be an issue for us either.

**Comment by calvincestari** [comment]
> If a key feature of the 2.0 upgrade is Swift 6 and Swift concurrency support, I don't think that should be a problem in general. I don't know of any iOS developers who rely on these technologies and still support older versions of iOS.

Our original deployment target for this work was iOS 13 where a lot of the concurrency API was back-ported to, but working through parts of the design yesterday showed that some of the API we want to use is only available on iOS 15. That's the reason for the new minimum deployment target. I share the same sentiment though that the vast majority of apps that were updated to use Swift Concurrency went directly to 15 or higher.

**Comment by jimisaacs** [comment]
Not a problem

**Comment by chasemac** [comment]
My company's min target is 17 soon to be 18 because we have to keep up with the latest security updates from apple. Please deploy this soon!

**Comment by productdevbook** [comment]
> Looking for feedback on another possible change here. We are strongly considering making Apollo iOS 2.0 target a minimum version of iOS 15. Would that be a problem for many users?

iOS 16. I think it would be ideal.

**Comment by jimisaacs** [comment]
I have my code compiling in Swift 6 language mode with a lot of `@preconcurrency`, but also quite a bit of intermediary `@unchecked Sendable` types. It would be very helpful to include the following changes ASAP, without waiting on major refactors for them:
- Add `Sendable` to `Cancellable` protocol
- Add `Sendable` to `GraphQLOperation` protocol
- Add `Sendable` to `RequestContext` protocol
- Add `@Sendable` to all public facing completionHandler and resultHandler blocks
- Add `Sendable` to `RequestChain` protocol

Low priority as I can add `Sendable` with `@retroactive`:
- Add `Sendable` to `CachePolicy` enum
- Add `Sendable` to `GraphQLResult` struct

It doesn't seem like I need much else than this at the moment.

**Comment by Brett-Best** [comment]
@AnthonyMDev iOS 15 is fine for us too, I was hoping we could get an update on the timeline that you're expecting for version 2. I understand Swift 6 language mode is complex, however, even importing Apollo with pre concurrency still presents problematic. I was wondering if perhaps some of the sendable improvements could be made in a version 1.x release?

**Comment by calvincestari** [comment]
@Brett-Best - @AnthonyMDev and I spoke about this earlier in the week and I believe we were going to look into some of the types that @jimisaacs requested above. The problem is that many of the types we've looked into before seem innocuous and simple but end up a sprawling complicated mess when you being changing them. We're still looking into it for now though.

**Comment by fabioferrero** [comment]
@calvincestari @AnthonyMDev do you have a conservative yet realistic timeline about this proposal? I would love to understand when during this year I would be able to work on this with my team.

**Comment by MarcoDotIO** [comment]
Hello there!

We are currently working on an SDK that requires the use of Apollo iOS. It would be great to hear any updates regarding this overhaul.

**Comment by CarlosdelamoraCare** [comment]
I would like to also have an API that doesn't return a stream, but throws and returns the data. I understand the use case of this 
>Because GraphQL requests may return results multiple times, the request methods will return an AsyncThrowingStream.

```
public func fetch<Query: GraphQLQuery>(
    query: Query,
    cachePolicy: CachePolicy = .default,
    context: (any RequestContext)? = nil
) -> AsyncThrowingStream<GraphQLResult<Query.Data>, any Error>
```

So it makes sense to keep it. But there is a very common use case that I believe deserves its own API. That is when you only return a single element.  Hence an API that looks like this would be great 
```
public func fetch<Query: GraphQLQuery>(
    query: Query,
    cachePolicy: CachePolicy = .default,
    context: (any RequestContext)? = nil
) -> async throws GraphQLResult<Query.Data>
```

**Comment by Vincz** [comment]
@calvincestari @AnthonyMDev Hi guys, any update regarding the project? It seems there is almost no activity in the past months and no news about the upcoming version. Do you mind to share a little update? Thanks!

**Comment by AnthonyMDev** [comment]
I'm sorry we haven't given an update in so long on this. I've been really stuck on an issue, but was finally able to find a way forward last week.

Apollo iOS has relied on some of the ["magic"](https://forums.swift.org/t/anyhashable-and-implicit-type-conversions/55728/2) of `AnyHashable` to allow the coercion of dynamic JSON values to the correct types at runtime. But because `AnyHashable` is explicitly marked as unable to conform to `Sendable`, it was incompatible with the newly updated APIs to work with the Swift 6 Concurrency Model. I tried many iterations of different solutions, all of which led to dead ends for quite a while. Many of these iterations required me to rethink the API design of the library, so they weren't just simple things I could try in a day. I've finally settled on a solution that, while not ideal, at least allows us to move forward.

I am having to do a conversion of the JSON values to `AnyHashable` internally, and then coerce that value back to `[String: any Sendable & Hashable]` before exposing it to the public APIs. This incurs some performance overhead, but with some careful testing, I've been able to verify does convert the Objective-C types that are returned from `JSONSerialization` into native Swift types that can be passed around as `any Sendable & Hashable` values properly.

---

I am now working on updating the `ApolloStore` and `NormalizedCache` to support `async` data loading functions. The `ApolloStore` utilizes a [Reader/Writer pattern](https://github.com/eleev/swift-design-patterns/blob/main/Concurrency%20Design%20Patterns/Barrier/Barrier.md) which is currently implemented using a GCD barrier queue. The Swift 6 concurrency model doesn't actually include any native way to support this. Actors cannot be used with multiple concurrent reads in any safe way. I've been doing a lot of research to figure out how to implement this, and again -- have had to go through multiple iterations to find the right approach.

I am currently exploring utilizing the [`SwiftAtomics`](https://github.com/apple/swift-atomics) package to write my own low-level Reader/Writer implementation. This is complex and very error prone, but I am optimistic that this solution will pan out.

---

I am truly disappointed that it is taking as long as it is to get something out to the community here. This entire project has been one seemingly unsolvable problem after another, and each problem has ended up affecting the solutions to the previous problems I thought I had solved. I feel like I'm fighting the [Hydra](https://tenor.com/bkEYS.gif) here.

But there is at least _some_ good news. For the first time in many months, I have been able to get a version of the project using the Swift 6 language model to compile successfully **and pass all of our existing unit tests!**

---

After some of the feedback on this issue, I have decided to refocus my efforts on getting a minimal viable changeset that will work with the Swift 6 language model distributed to hopefully unblock those of you who need this. It will still be an API breaking change, but should be as unobtrusive as possible for users.

This will not take full advantage of the new concurrency model, and will not be the final product for 2.0. Instead it will be a "1.x + Swift 6 support" variation of the existing project. Once I have that published for all of you to use, I will continue on with the other API changes we want to make to fully take advantage of `async/await`, as well as other breaking changes that we know we want to make in order to support future planned feature work. That will then eventually be released as Apollo iOS 2.0.

I hope that providing this variation of the library will enable you all to do the work you want to do with Apollo iOS. That will give us the time we need to make 2.0 what we really want it to be and prevent too many additional breaking changes once 2.0 is released (we would like to avoid releasing a 2.0 and 3.0 version within 6 months of each other).

---

In the mean time, my colleague @BobaFetters has just put up an [RFC for the early imaginings of reworking our caching mechanisms](https://github.com/apollographql/apollo-ios-dev/pull/615) to provide a ton of those new features I alluded to up above. That work is only just beginning, but it is being done in parallel to this project. Please take a look at that and give us your comments and opinions on it as well.

You all have my sincere apologies for the time this project has been taking. No one is more frustrated or disappointed that we haven't solved this yet than I am myself. Thank you for your patience with me.

**Comment by Brett-Best** [comment]
@AnthonyMDev thanks for the reply and update, without knowing the details too much I just wanted to share https://github.com/pointfreeco/swift-concurrency-extras/blob/main/Sources/ConcurrencyExtras/AnyHashableSendable.swift which you've likely seen already but just in case you haven't I'm sharing it :)

**Comment by AnthonyMDev** [comment]
@Brett-Best thanks for sharing! I did already look at that one, and it sadly won't work for us here. The actual `AnyHashable` type has some special "magic" that it does for bridging and conversion of types under the hood. This type-erased wrapper doesn't do that magic, which is understandable, since it would be impossible to do without baking the type into the language itself.

**Comment by postmechanical** [comment]
Don't burden `NormalizedCache` with `AnyActor` just make it `Sendable` and let implementers decide whether an `actor` is right for their use case. `NormalizedCache` needs to support expiration. There needs to be both `async` single result returning query methods for fetches and cache policies that will only ever return a single result and those that use `AsyncThrowingStream` for fetches and cache policies that may return more than a single result over time. I know all these points have already been made but just wanted to reiterate them here as they are important for my use cases.

**Comment by AnthonyMDev** [comment]
Thanks for the feedback @postmechanical! I've been working through the design of the `ApolloStore` and `NormalizedCache` for 2.0 recently. I'll update the RFC with that new info soon, but here's a summary of what I've settled on.

We're going to continue to handle locking read/write access to the `NormalizedCache` at the `ApolloStore` level. That means that, as long as you only use your cache with a single `ApolloStore` and don't do any other external read/write access that doesn't go through the `ApolloStore` you shouldn't have to worry about concurrency at all. I'm still experimenting, but I don't think we even need to make the cache `Sendable`.

For 2.0, we've decided to cut the scope down to just what is necessary to get `async/await` working so that we can get something in the hands of users as soon as possible. I still plan on providing methods for single fetches and for cache policies with multiple results that use an `AsyncThrowingStream` in 2.0. 

But support for expiration/TTL will be coming in a future version (maybe a 2.x, but more likely a 3.0) that we are targeting to release later this year. @BobaFetters is leading a complete rewrite of the underlying cache mechanisms that will support this as well as other big feature additions to the cache. He has an [RFC up here](https://github.com/apollographql/apollo-ios-dev/pull/615) that you can check out for that work.

**Comment by AnthonyMDev** [comment]
Hi everyone! I've got a big update for you all finally! I now have a preview branch of a lot of the 2.0 changes ready for you to take a look at. There is still a lot of work to be done, but I've sorted out enough of the foundational logic to feel confident in the direction we are taking here.

The `ApolloClient` APIs for fetching data are not updated yet, so those are still using the completion handler syntax for now. But the `NetworkTransport` and `RequestChain` APIs are all new and fully taking advantage of the Swift Concurrency Model. A lot of the other public data structures and API surface are now using `Sendable` and `async/await`.

The pieces that have had the most dramatic changes are in the `RequestChain` and the Interceptors. If you were using a custom interceptor chain, you're going to need to re-write your interceptors. But once you do, I think you'll be happy with the new way of doing things.

There are a lot of rough edges, so I'd be surprised if anyone is able to get this working well with their existing codebase yet, but if you'd like to take a look at what we've got, you can check out the `v2-prerelease` branch of the `apollo-ios-dev` repo.

I'm going to be on vacation next week, but when I get back, I plan on tidying a few things up and getting the `ApolloClient` APIs to use `async/await`. Once I've got that done, I'll be releasing an Alpha release that is ready for public testing.

**Comment by AnthonyMDev** [comment]
Happy to provide some examples and a bit more context!

Here is the interface for the interceptors for 2.0:

```swift
public protocol ApolloInterceptor: Sendable {

  typealias NextInterceptorFunction<Request: GraphQLRequest> = @Sendable (Request) async throws -> InterceptorResultStream<Request.Operation>

  func intercept<Request: GraphQLRequest>(
    request: Request,
    next: NextInterceptorFunction<Request>
  ) async throws -> InterceptorResultStream<Request.Operation>

}
```

There are a few fundamental changes to the interceptor and request chain pattern that you need to be aware of in order to make sense of this stuff. I'll be getting more documentation of all of this before we go to Beta, but for now I'll summarize here.

1. Interceptor's now will only receive the `request` initially. You can modify the request and then pass it to the `next(request)` block. The `next(request)` block returns a `InterceptorResultStream` which you can then call `map` on to transform the output response objects. This means that it is clear when you have access to the response, instead of it being an optional parameter and needing to make sure that you place your response handling interceptors after the interceptor that handles network fetches. Also, because you are working with a stream with a `map` function, it is more clear that this `map` block may be called multiple times (for multipart responses to support defer and subscriptions).
2. There is no need for a `NetworkFetchInterceptor` anymore. You can provide a `URLSession` object, and when the request chain gets to the last interceptor, it will handle sending a network request with the given `URLSession`. You still do need to provide the `JSONResponseParsingInterceptor` somewhere in your chain to parse the network response into a parsed `GraphQLResult` though.
3. Interceptors are now traversed down the stack for modifying the request _and then back up the stack with the response_. _This means you'll need to adjust the order of your interceptors._ It's no longer just a linear list of actions.
4. Cache interceptors are separate from the request interceptors. They have a different API with functions for reads and writes. And now you only need to provide one that will have its read/write functions called at appropriate times by the `RequestChain`. Because we now have `CacheInterceptor`, I'm considering renaming `ApolloInterceptor` to `RequestInterceptor` for clarity.

Now for some examples:

A logging interceptor can be implemented like this:

```swift
struct LoggingInterceptor: ApolloInterceptor {
  func intercept<Request: GraphQLRequest>(
    request: Request,
    next: NextInterceptorFunction<Request>
  ) async throws -> InterceptorResultStream<Request.Operation> {
      print("Starting \(request.operation.operationName)")
        
      do {
        return try await next(request).map { response in
            print("Success \(request.operation.operationName)")
        }

      } catch {
        print("Failure \(request.operation.operationName) with error: \(error)")
        throw error
      }      
  }
}
```

An authorization interceptor might look like this:

```swift
struct AuthorizationInterceptor: ApolloInterceptor {
  func intercept<Request: GraphQLRequest>(
    request: Request,
    next: NextInterceptorFunction<Request>
  ) async throws -> InterceptorResultStream<Request.Operation> {
      var request = request
      request.addHeader(name: "Authorization:, value: "MyAuthKey")
       
      // This interceptor doesn't need to operate on the response, so it can just pass the stream
      // returned by the `next` block on to the next interceptor.
      return try await next(request)
  }
}
```

**Comment by jimisaacs** [comment]
> There is no need for a NetworkFetchInterceptor anymore. You can provide a URLSession object, and when the request chain gets to the last interceptor, it will handle sending a network request with the given URLSession. You still do need to provide the JSONResponseParsingInterceptor somewhere in your chain to parse the network response into a parsed GraphQLResult though.

This is a rather alarming change. We don't pass around URLSessions for network access, and depend on an interceptor for using our networking. How should this be handled?

**Comment by calvincestari** [comment]
@jimisaacs - one of the biggest pieces of feedback we got was that it's difficult to use a custom network stack with Apollo iOS. This change is to address that and provide for more flexibility with `URLSession` and responding to it's events. With the networking calls being pulled internally it's not going to be possible to do as before via the interceptor. What did your changes to the `NetworkFetchInterceptor` look like?

@JOyo246 - Anthony will be able to give a better answer but I believe the reason those changes are scoped to 2.0 is exactly that; they sprawl out so much that it's not possible to have a non-breaking change for 1.x that includes those.

**Comment by jimisaacs** [comment]
@calvincestari 
> What did your changes to the NetworkFetchInterceptor look like?
There were no changes to `NetworkFetchInterceptor` we do not use it. We have our own interceptor that does it's own networking entirely.

**Comment by jimisaacs** [comment]
>  it's difficult to use a custom network stack with Apollo iOS

We are and have used a completely custom networking stack since we started using Apollo, though it has never utilized `NetworkFetchInterceptor` to do so.

**Comment by AnthonyMDev** [comment]
@jimisaacs  I'm on a cruise, so can't give super detailed answers right now, but I can respond quickly!

You just need to conform to the new ApolloURLSession protocol. So it doesn't even have to be a URLSession object exactly. Just something that conforms to the protocol we've defined. We just defined the protocol so that URLSession does conform to it by default!

You provide an object that conforms to the ApolloURLSession protocol through your InterceptorProvider.

**Comment by AnthonyMDev** [comment]
As I get closer to finishing the API's for the client, I'm strongly considering another design change that I'd like to get some feedback on.

I'd like to move the reading and writing of data out of the `RequestChain` completely. Having the `ApolloClient` handle reading prior to kicking off a request chain and the writing after it gets results. This would mean that the `NetworkTransport` (and consequently `RequestChainNetworkTransport`) would only be responsible for executing a network request and parsing the response data (plus any middleware you want to add to the `RequestChain`).

I understand that there are some advanced uses cases for custom cache interceptors out there (particularly I'm aware of @jimisaacs needing these). While we are planning the 3.0 overhaul of the caching mechanisms to provide TTL support and other much requested features, I still want to support the solutions to implementing these features that users have come up with in 2.0.

The thought is to have a `CacheInterceptor` protocol that looks like this:
```swift
public protocol CacheInterceptor: Sendable {

  func readCacheData<Query: GraphQLQuery>(
    for query: Query,
    from store: ApolloStore
  ) async throws -> GraphQLResult<Query.Data>

  func writeCacheData<Operation: GraphQLOperation>(
    cacheRecords: RecordSet,
    for operation: Operation,
    with result: GraphQLResult<Operation.Data>,
    to store: ApolloStore
  ) async throws

}
```

The `CacheInterceptor` would be passed to the `ApolloClient` on initialization, instead of into the `RequestChain` via the `InterceptorProvider`. Would this still serve your use cases?

**Comment by bainfu** [comment]
@AnthonyMDev I think that would be huge to debug some of our queries, and whether or not some of the mutations are writing to where we think it would be.

**Comment by AnthonyMDev** [comment]
Update: There is a [new PR](https://github.com/apollographql/apollo-ios-dev/pull/678) up for 2.0 that includes the (hopefully) finalized APIs of the `RequestChain`.

Documentation and code organization for this code still needs to be done, but this PR provides a deep look into what I've been working on in the past few weeks.

The PR has a long description that discusses the functional changes. I'm not expecting anyone to read all of the code in the PR (except @calvincestari and @BobaFetters 😉), but I would love interested users to read the PR description and give feedback!

**Comment by AnthonyMDev** [comment]
Just wanted to give you all an early heads up. I will be publishing the first official preview release of Apollo iOS 2.0 tomorrow. We've still got a lot of work to do, but we're at a point where things are building and we are ready for some feedback on how it works. Though we have a lot of unit tests, we fully expect there to be some bugs right now.

There will be more information when I publish the release tomorrow, but if you want to start looking at the project now, it's is up on the `v2-alpha` branch of our `apollo-ios-dev` repo. The release tomorrow will include a branch pushed to the main `apollo-ios` repo.

**Comment by markusfassbender** [comment]
Basically you can have the old and the new Apollo client with different versions in your app by making your own package with a wrapper. If you have naming conflicts you can change the names or make a simpel fork of the old version.

Then make the credential store shared and you are free to move incrementally to version 2.0 👍

**Comment by AnthonyMDev** [comment]
Thank you all for the feedback on maintaining the old APIs. I'm open to keeping the core `ApolloClient` APIs for `fetch(query:)`, `watch(query:)`, `perform(mutation:)`, and `subscribe(subscription:)` around for 2.0. Maintaining those shouldn't be too difficult. I may look into exposing them as a separate target within the Apollo iOS package that you can optionally include for legacy API support. 

I expect these to cover most of the surface area in your applications, so this should make the migration a lot easier.

However there are a significant number of other core APIs that we aren't going to be able to keep the old versions around for. The `ApolloClient` initialization and configuration options have changed. The `NormalizedCache` APIs are now `async`, so if you have a custom cache implementation, you will need to update it when migrating. The interceptors for custom networking through the Request Chain have changed significantly as well. There are a number of other things that have changed which will require some changes on your part, but these are parts of the library where you are usually doing configuration in a single place.

**Comment by martin-muller** [comment]
Cool to see async supported at the fundamental level, nice one! I've spent some time migrating our project and below are some notes I have:

1. Generated `LocalCacheMutations` don't compile
    This seems to be a bug in the first alpha, the template defines the type for local cache mutation as struct https://github.com/apollographql/apollo-ios-codegen/blob/2.0.0-alpha-1/Sources/ApolloCodegenLib/Templates/LocalCacheMutationDefinitionTemplate.swift#L23, but `LocalCacheMutation` protocol  conforms to `AnyObject` https://github.com/apollographql/apollo-ios/blob/2.0.0-alpha-1/Sources/ApolloAPI/LocalCacheMutation.swift#L1 and thus requires a reference type.

2. I couldn't find overload for `watch` that would return an async stream. We've been using an overload like that in our wrapper but that's without the fundamental async support, do you think it could make its way into Apollo itself?

3. Custom network mocking
    We have our custom mocking layer that pre-dates the mocking infrastructure that Apollo provides. We essentially take the production models we already have generated, initialise them, serialise them and pass them to the interceptor chain as a response. I was having issues getting that to work with the new interceptor structure, do you think I should be able to get that working with the new structure?

**Comment by AnthonyMDev** [comment]
Thanks so much @martin-muller! For trying out the alpha!

1. I'll take a look at this one. Should be an easy fix.

2. The reason why I didn't include an async stream API for query watcher is because they aren't terminal on their own, so your `for/in` loop would run forever. Maybe I'm thinking about this wrong. I'd love some feedback from others here.

3. I think you should still be able to achieve something like this. A few ways I could see this working:

- Make the mocking interceptor your `CacheInterceptor` and just return the response there as if it was from a cache read. This won't work for requests that are using a cache policy which ignores the cache though. 
- Make the mocking interceptor the `ResponseParsingInterceptor`. You'll need a stubbed `ApolloURLSession` that just returns an empty, phony `HTTPURLResponse` and data stream, and then have your parsing interceptor completely ignore the response and just return your mocks there.

If you're not able to make that work, please provide me some more details on what is blocking you!

**Comment by AnthonyMDev** [comment]
> I don't know if it'd be possible, but could the Int32 only be applied to input types? That's the only part of the requests with a safety concern to prevent too big of a value being sent to a server.

That's a really interesting idea. I think that probably will work. Let me think through it a bit more.

**Comment by AnthonyMDev** [comment]
Wanted to give another update. I've been looking into the web socket implementation, and it's going to be a bit more involved than expected. In order to get 2.0 released, we've made the decision to ship the initial release without web socket support. We'll be prioritizing adding web sockets in a 2.x release soon (hopefully 2.1 will include it). By moving web sockets to a future release, we hope to have a stable version of 2.0 ready in the new few weeks!

You can still execute GraphQL Subscriptions with Apollo iOS 2.0 over HTTP multipart.

**Comment by AnthonyMDev** [comment]
@CraigSiemens If we made it so that `Int32` was only used for input types, would you still feel that we need to make it an optional feature? I'd really like to just have one opinionated way of doing this and not have to maintain additional configuration options. Making this configurable would be a deep-reaching change in the code generation engine.

I think you're right that the only place it really matters to be `Int32` is when we are sending it to the server. If servers send _us_ an `Int32`, we can still treat it as an `Int` and there are no issues. The only time this becomes an issue is if you were to increment that returned value (such as in a local cache mutation) to a number larger than `Int32.max` _and then also pass that value to an input object/parameter_. 

It's more likely that `Int` values being passed to a GraphQL input are coming from user input via UI or developer calculated. 

By generating GraphQL inputs as `Int32`, the developer is made to be aware that they need to convert their `Int` to `Int32` and consider the implications of integer overflow here. This prevents users from having a configuration option that gives them an additional foot-gun while also maintaining the usability of `Int` in all the places where it's not absolutely necessary to work with an `Int32`.

**Comment by AnthonyMDev** [comment]
@justnicke Bug fix [PR](https://github.com/apollographql/apollo-ios-dev/pull/730) for your issue is up. This will be included in the next Alpha/Beta release. Thanks so much for the report!

**Comment by ahtcx** [comment]
@ppmtrifork the `apollo-ios-cli.tar.gz` file is missing from the `beta-1` release assets, you may be using an older CLI version.

I'm using [mise](https://mise.jdx.dev/) to manage my tool versions, which previously had been working with the following configuration:

```toml
[tools]
"apollo-ios-cli" = "2.0.0-beta-1"
```

This broke though as it uses [mise-plugins/mise-apollo-ios-cli](https://github.com/mise-plugins/mise-apollo-ios-cli) which retrieves the CLI from the release assets.

Instead I've updated my tools to use the [HTTP Backend](https://mise.jdx.dev/dev-tools/backends/http.html) which works great:

```toml
[tools]
"http:apollo-ios-cli" = { version = "2.0.0-beta-1", url = "https://github.com/apollographql/apollo-ios/raw/refs/tags/2.0.0-beta-1/CLI/apollo-ios-cli.tar.gz" }
```

**Comment by AnthonyMDev** [comment]
Thank you for the bug reports everyone! I've attached the CLI binary to the `beta-1` release, so @ahtcx your issue should be resolved.

@ppmtrifork & @pixelmatrix I'm going to look into these issue today. I'll get a new beta with fixes out ASAP. 

It's so helpful to have you all trying out the beta and sending these reports. I really appreciate it!

**Comment by pixelmatrix** [comment]
Confirmed that I'm not getting errors with generated code anymore

**Comment by ppmtrifork** [comment]
> Beta 2 is up, would love to get confirmation that the issues you've experienced are resolved now!

I can confirm that the previous issue was resolved, just like @pixelmatrix . Thanks so much!

Nevertheless, a new issue appeared: in our implementation we conform `Date` to `CustomScalarType`. One of the protocols in `CustomScalarType` is `OutputTypeConvertible`. In Apollo 1.x the conformance to `OutputTypeConvertible` worked out of the box, but now I'm required to implement in manually. No problem in doing that, but when adding the code to conform to `OutputTypeConvertible` I get the following error:

> No type named 'Selection' in module 'ApolloAPI'

Any idea on what's happening here? Thanks in advance!

**Comment by Brett-Best** [comment]
Just wondering if you have plans for further betas or release candidates and if you have a date in mind for a stable release?

Out of curiosity would you currently recommend using beta 2 in production?

**Comment by AnthonyMDev** [comment]
Hi @Brett-Best, we haven't received any more bug reports to fix, so the beta feels pretty stable right now. We are working on a few feature additions, but those shouldn't be too much of a breaking change. The primary thing I'm focusing on right now is updating the documentation and creating usage guides and a migration guide. Once all of that is done, we'll be releasing a stable public release. Our target is the end of September for that release.

I don't anticipate using the beta 2 to cause you any significant problems, but there may be a few minor breaking changes you would need to deal with again when updating to the public release. More importantly, it may be tough to upgrade an entire 1.0 application to 2.0 without the migration guide being published.

**Comment by ScottPlease** [comment]
Is there an estimate for the `WebSocketTransport` implementation? In the current v2 branch it [says](https://github.com/apollographql/apollo-ios/blob/v2/Sources/ApolloWebSocket/WebSocketTransport.swift) 

> WebSocketTransport has not yet been implemented for Apollo iOS 2.0. This will be implemented prior to beta release.

**Comment by ppmtrifork** [comment]
@AnthonyMDev Just a minor hiccup: in the generated file that contains the schema I get the following _warning_

> Static property 'configuration' is not concurrency-safe because non-'Sendable' type 'any SchemaConfiguration.Type' may have shared mutable state

This happens in the line with the `static` constant within the `SchemaMetadata` enum:

```swift
public enum SchemaMetadata: ApolloAPI.SchemaMetadata {
  public static let configuration: any ApolloAPI.SchemaConfiguration.Type = SchemaConfiguration.self
  @_spi(Execution) public static func objectType(forTypename typename: String) -> ApolloAPI.Object? {
  ...
  }
}
```

It's just a warning, but reporting in case it becomes an error in the future.

**Comment by AnthonyMDev** [comment]
@ScottPlease I **just** discovered that the `RequestChainNetworkTransport` does not conform to `SubscriptionNetworkTransport` a few hours before receiving this message. 😆

All of the inner workings to support HTTP subscriptions are already in place and are unit tested. I just forgot to hook them up to the public API. This will be done in the next beta release, which is coming Monday!

---
## Progress Update

I've been working on updating all of the inline documentation this week. That is complete and will be available with the beta 3 on Monday. If you see any issues with the documentation in beta 3, please file an issue.

Next week, I will be starting on the usage guides and documentation for the Apollo Docs site. This will include a detailed migration guide.

**Comment by ScottPlease** [comment]
Hey @AnthonyMDev , any updates on the aforementioned beta 3? Excited to try out the subscriptions 😄

**Comment by AnthonyMDev** [comment]
@PatrickDanino sounds like you're understanding it correctly!

> As part of our existing code, we also need to have access to the response headers so we can correlate the error to the a specific response in our logs. How would we do that in v2?

The problem is that an error that occurred pre-flight won't have a response, so you'll need to ensure the response is provided to your interceptor. There are two ways I could see you doing that.

1. Make sure that any error thrown by other post-flight steps of your `RequestChain` include the response as a property of the error. You could define a `protocol ResponseProvidingError` for your errors to conform to here. But you're now responsible for making sure all your errors include the response. If this logic is only necessary for specific types of errors, this could be useful though.

2. Create an `HTTPInterceptor` that provides the response headers you are looking for as a `TaskLocal` value. The usage docs will have a section on how to do this when they are published. Then your error handling interceptor can access that `TaskLocal` value to get the info it needs. This method makes sure that the response headers are present for the error handler even if the error occurs during the parsing of the response JSON body and no matter what type of error is thrown.

**Comment by PatrickDanino** [comment]
Both of these options sound non-obvious to me.

I understand that not all errors will have a response, which is from my recollection why it was optional in V1, but that sounds like a lot of hoops to have to jump through just to get the response (if there is one).

I also get the intent around wanting to abstract this complexity away from clients, but we can't be the only ones with a need to access the underlying response, so for those of us who need it, this adds complexity compared to V1.

**Comment by AnthonyMDev** [comment]
@PatrickDanino I've spent a lot of time thinking about this since yesterday! There isn't really a good way to provide the response as part of the stream without severely complicating the API. 

A big part of the issue here is that what you are doing crosses the separation of responsibilities between the `HTTPInterceptors` and the `GraphQLInterceptors`. Exposing the entire `HTTPURLResponse` to the `GraphQLInterceptors` would make the `GraphQLInterceptors` dependent on the HTTP networking layer. This would be a problem, because I actually plan on building a version of a Request Chain for Web Sockets. It would use the same `GraphQLInterceptors` with `WebSocketInterceptors` instead of `HTTPInterceptors`.

But really, it sounds like in this situation, you don't need the entire `HTTPURLResponse`, you actually only need the value of one or two particular fields from the response header that represent some data your application needs. I think this is a perfect example of the intended use of an `HTTPInterceptor` and `TaskLocal` values. 

You create a custom `HTTPInterceptor` that extracts only the needed values from the `HTTPURLResponse` and creates some strongly typed data structure that your other interceptors need to consume for their business logic. In 1.0, you would then be able to add that data structure to the `RequestContext` to pass it to subsequent interceptors. In 2.0, we replaced `RequestContext` with use of `TaskLocal`. This seems non-obvious right now of course, but it will be highlighted prominently in the usage examples and documentation for the new `RequestChain` when that's all published.

This keeps the separation clear between the GraphQL layer and the HTTP layer.

**Comment by andykent** [comment]
Just highlighting that we also have this wish...

> I couldn't find overload for watch that would return an async stream. We've been using an overload like that in our wrapper but that's without the fundamental async support, do you think it could make its way into Apollo itself?

@AnthonyMDev earlier in this thread you said:

> I'm not sure I'm happy with the state of GraphQLQueryWatcher in Alpha 1. It's the only place where we are still using a closure based result handler instead of an AsyncStream. My initial thought was that, because a watcher never ends unless canceled by the user, an await for result in watcher on the AsyncStream would be a loop that never terminates on its own. I'm not sure how clear that is to the user.

Am I missing something obvious here or is the correct fix not to have `watch` participate in Task cancellation? In this situation it would need to return an `AsyncThrowingStream` that can throw a CancellationError. Consumers could then do:

```swift
let watcher = Task { 
  for try await data in watch(query: MyQuery()) {
    ...
  }
}

// Sometime later...
watcher.cancel()
```

A really big advantage of this is it plays nicely with the SwiftUI task modifier. We often find ourselves wanting a query watched for the duration of a views lifecycle. Using this approach means you don't need to worry about cancellation at all.

```swift
struct Example: View {
  @State private var message = "loading"
  
  var body: some View {
    Text(message)
    .task {
      for try await data in watch(query: MyQuery()) {
        switch result {
          case let .success(data): message = data?.message ?? "missing"
          case let .failure(error): message = error.localizedDescription
      }
    }
  }
}
```

The only downside here I can see if that `watch` would return a throwing stream but would never be expected to throw anything that wasn't a CancellationError so it might be slightly annoying to handle if you know you never intend to cancel the task (seems uncommon?). I think this is still more correct and much safer though, otherwise in the more common scenario of needing to cancel a running stream you have to remember to cancel both the `Task` and the underlying `GraphQLQueryWatcher` which feels a lot more error prone.

**Comment by andykent** [comment]
No problem, really glad my comment was helpful! 😄

After having a couple of days to think on this I'm pretty sure that the stream doesn't need to be throwing. I think it can be done with an AsyncStream as long as it responds to task cancellation correctly by terminating the stream it should "just work" I think.

The other thing I realised from my above example is that I had assumed we would only need to return the stream but actually I think we do need to return a tuple still to maintain support for manually calling `refetch()` (we do use this functionality quite a bit). Although then that raises the question of if the other thing returned should still be `GraphQLQueryWatcher` or not. Obviously the simplist change is to just return it as currently but then is it confusing that this also has a `cancel()` function if it's usually not needed?

The other options would be to just provide a refetch function or even to provide some specialisation for AsyncStream that adds a `refetch()` function directly on the stream, that might be nice but I'm not sure if that's possible.

Putting all that together we might end up with something that could be used along the lines of:

```swift
struct Example: View {
  @State private var message = "loading"
  @State private var refetchAction = () -> nil
  
  var body: some View {
    VStack {
      Text(message)
      Button("Force Refetch", action: refetchAction)
    }
    .task {
      let (results, refetchA) = await watch(query: MyQuery())
      self.refetchAction = refetch
      
      for await data in results {
        switch result {
          case let .success(data): message = data?.message ?? "missing"
          case let .failure(error): message = error.localizedDescription
      }
    }
  }
}
```

This is a pretty similar API to what we currently have in our own wrapper around the 1.0 API which looks something like the following:

```swift
func watch<Query: GraphQLQuery>(
  query: Query,
  cachePolicy: CachePolicy = .returnCacheDataElseFetch,
  queue: DispatchQueue = .main
) -> (GraphQLQueryWatcher<Query>, AsyncStream<Result<GraphQLResult<Query.Data>, Error>>) {
  var continuation: AsyncStream<Result<GraphQLResult<Query.Data>, Error>>.Continuation?
  let stream = AsyncStream { c in
    continuation = c
  }
  let watcher = watch(
    query: query,
    cachePolicy: cachePolicy,
    refetchOnFailedUpdates: true,
    callbackQueue: queue
  ) { result in
    continuation?.yield(result)
  }
  continuation?.onTermination = { @Sendable _ in watcher.cancel() }
  return (watcher, stream)
}
```

And in usage looks like:

```swift
    let (watcher, results) = PortalAPI.apollo.watch(query: MyQuery())
    self.watcher = watcher

    for await result in results {
      ...
    }
```

I may have time to take a look at a PR for something along these lines on Monday but I will need to see how my schedule is looking so no promises.

A couple of questions:

1. am I correct in assuming that all the other APIs in the v2 branch should be good task cancellation citizens, e.g. cancelling in-flight network requests, skipping interceptor processing on cancellation etc? (If not this feels like a bug)

2. does the stream returned from `CachePolicy.Query.CacheAndNetwork` work like this already. e.g. if the task is cancelled before it goes to the network is the expectation that the stream only returns 1 value and then terminates? (If not this feels like a bug)

3. do you have opinions on if we should return `(AsyncStream, GraphQLQueryWatcher)` or some other more limited refresh mechanism? Possibly a struct with `callAsFunction` `(AsyncStream, RefreshAction)`? Maybe there is a better API here?

**Comment by ScottPlease** [comment]
Sorry false alarm, SPM was misbehaving. It wasn't resolving to the correct commit with the branch policy. Just needed to get Xcode to behave properly.
